### PR TITLE
Ensure IterableCoder handles the stream which is under its responsability as expected + add utility SkipCloseCoder

### PR DESCRIPTION
**Pull request by rmannibucau**
DESCRIPTION HERE

IterableCoder virtually write N times using a delegate coder (element coder). However it doesnt ensure the delegate coder reads only the part of the stream it writes so it must be fixed to enforce its responsability.

This PR also adds a SkipCloseCoder utility to avoid to do it in several coder implementations.

Finally it adds a Delegating API to be able to unwrap coders to validate them - coder generics are not handled in current code on one side and on another side several casts enforces this kind of API to enable composition of features in coders.

WIP: the coder serialization changes so it needs love for other languages than pythons and standard_coders.yml

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand:
   - [ ] What the pull request does
   - [ ] Why it does it
   - [ ] How it does it
   - [ ] Why this approach
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

### CODE REVIEW

**Comment by jkff** [comment]
Seems the consensus on the dev@ thread was that this PR in the current form is not needed, or at least that Beam is not ready to accommodate it until the next major version? Should it be closed?

**Comment by rmannibucau** [comment]
Not sure. Most of it is still useful but can need to have a flag on custom coder to activate it like @WrapWithLength - name is bad.

List outcome is more yes it needs to be fixed but we cant break current serialization.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/LengthAwareCoder.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+import org.apache.beam.sdk.util.LimitedInputStream;
+import org.apache.beam.sdk.util.VarInt;
+import org.apache.beam.sdk.values.TypeDescriptor;
+
+/**
+ * A coder encoding in memory and marking the size of the encoded value
+ * to let be decoded enforcing this size.
+ * @param <T> the supported type of the instance which can be encoded/decoded.
+ */
+public class LengthAwareCoder<T> extends Coder<T> implements Delegating<Coder<T>> {
```
[the remark this answers is not in the document]
Comment (rmannibucau) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/coders/LengthAwareCoder.java]: Right, it just supports delegation properly.

**Comment by rmannibucau** [comment]
Filter*Stream are vicious cause buggy on some systems and a source of bottleneck is not fully delegated/reimplemented. Not the best we can offer to the users OOTB.

**Comment by rmannibucau** [comment]
replaced the custom lengthaware coder by the existing one (prefix)

**Pull request closed without merging**